src/App/Fossa/YamlDeps.hs

```diff
@@ -37,19 +36,16 @@ import Path
 import Srclib.Converter (depTypeToFetcher)
 import Srclib.Types (AdditionalDepData (..), Locator (..), SourceUnit (..), SourceUnitBuild (..), SourceUnitDependency (SourceUnitDependency), SourceUserDefDep (..))
 
-analyzeFossaDepsYaml :: (Has Diagnostics sig m, Has ReadFS sig m, Has (Lift IO) sig m) => Path Abs Dir -> Maybe ApiOpts -> m [SourceUnit]
+analyzeFossaDepsYaml :: (Has Diagnostics sig m, Has ReadFS sig m, Has (Lift IO) sig m) => Path Abs Dir -> Maybe ApiOpts -> m (Maybe SourceUnit)
 analyzeFossaDepsYaml root maybeApiOpts = do
   maybeDepsFile <- findFossaDepsFile root
   case maybeDepsFile of
-    Nothing -> pure []
-    -- If the file exists and we have no SourceUnit to report, that's a failure
+    Nothing -> pure Nothing
     Just depsFile -> do
       yamldeps <- context "Reading fossa-deps file" $ readContentsYaml depsFile
-      sourceUnit <- context "Converting fossa-deps to partial API payload" $ Just <$> toSourceUnit root yamldeps
-      archiveSrcUnit <- case maybeApiOpts of
-        Nothing -> pure $ archiveNoUploadSourceUnit (vendoredDependencies yamldeps)
-        Just apiOpts -> archiveUploadSourceUnit root apiOpts (vendoredDependencies yamldeps)
-      pure $ catMaybes [sourceUnit, archiveSrcUnit]
+      -- If the file exists and we have no dependencies to report, that's a failure.
+      when (hasNoDeps yamldeps) $ fatalText "No dependencies found in fossa-deps file"
```

Comment:
[Important]: This line should be a part of `toSourceUnit`.  That constructor should take the sole responsibility of deciding whether we have enough information to construct a valid source unit.

Reply:
I don't know if I agree with this. `toSourceUnit` is converting a list of locators into a sourceUnit. It feels wrong to allow that constructor to exit 1 the entire program. IMO the caller should be the one to decide whether the output is acceptable or not.

Reply:
It doesn't fail the program, it just fails the `Diagnostic` carrier.  In this case, we intentionally let that diagnostic bubble up all the way to the main `analyze` handler.  `toSourceUnit` is a smart constructor, and it's only responsibility is to transform some inputs into a source unit, which is not guaranteed to succeed.  In this codebase, `fatal`, and its `Text`-specialized version `fatalText` will "throw a diagnostic error".  Which can be caught or left to flow upwards.

Reply:
I'm still not sure I agree with the idea that the constructor should be allowed to fail for the reason we are failing it, but I'm comfortable changing this to preserve the previous behavior.